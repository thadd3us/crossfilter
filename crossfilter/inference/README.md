# Crossfilter Inference Tools

This directory contains tools for computing embeddings and running inference on images using various models.

## compute_embeddings_cli.py

A command-line interface for computing embeddings for images and storing them in a SQLite database. The CLI supports incremental computation, UMAP projection, and parallel processing.

### Features

- **Incremental Processing**: Only computes embeddings for images that don't already exist in the database
- **Batch Processing**: Processes images in configurable batches for efficient memory usage
- **Parallel Database Writes**: Uses a background worker thread to write embeddings to the database in parallel with computation
- **UMAP Projection**: Optionally projects embeddings to 2D space using UMAP with spherical geometry
- **Progress Tracking**: Shows progress bars for long-running operations
- **SQLAlchemy Integration**: Uses SQLAlchemy for robust database operations

### Usage

```bash
python -m crossfilter.inference.compute_embeddings_cli \
    --embedding_type SIGLIP2 \
    --input_dir /path/to/images_by_uuid \
    --output_embeddings_db /path/to/output.db \
    --batch_size 10 \
    --recompute_existing_embeddings \
    --reproject_umap_embeddings
```

### Output Database Schema

The CLI creates the following tables:

#### `{EMBEDDING_TYPE}_EMBEDDINGS`
- `UUID` (TEXT PRIMARY KEY): The UUID of the image
- `EMBEDDING` (BLOB): The embedding vector serialized with msgpack

#### `{EMBEDDING_TYPE}_UMAP_MODEL`
- `MODEL` (BLOB): The UMAP model serialized with pickle

