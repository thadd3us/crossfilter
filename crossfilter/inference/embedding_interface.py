"""
Common interface for embedding computation classes.

This module defines the abstract base class that all embedding implementations
should inherit from, ensuring a consistent API across different embedding types.
"""

from abc import ABC, abstractmethod
from pathlib import Path

import numpy as np


class EmbeddingInterface(ABC):
    """Abstract base class for embedding computation."""

    @abstractmethod
    def compute_image_embeddings(self, image_paths: list[Path]) -> list[np.ndarray]:
        """
        Compute embeddings for a list of image paths.

        THAD: TODO: No parallel arrays -- take a DataFrame as input, add a column containing the embeddings.  Don't return anything.

        Raises:
            FileNotFoundError: If any image path doesn't exist
            ValueError: If any image cannot be loaded
        """
        pass

    @abstractmethod
    def compute_text_embeddings(self, captions: list[str]) -> list[np.ndarray]:
        """
        Compute embeddings for a list of text captions.

        Args:
            captions: List of text strings to encode

        Returns:
            List of 1D numpy arrays, one embedding vector per input caption
        """
        pass
