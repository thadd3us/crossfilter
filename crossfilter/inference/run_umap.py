"""
UMAP projection functions for SigLIP2 embeddings with spherical output space.

This module provides functions to run 2D UMAP projections on SigLIP2 embeddings
using spherical embedding space with Haversine distance in the output metric.
This produces coordinates that naturally lie on a sphere, perfect for geographic-style
visualization and analysis.

See: https://umap-learn.readthedocs.io/en/latest/embedding_space.html#spherical-embeddings
"""

import logging

import numpy as np
import pandas as pd

from crossfilter.core.schema import SchemaColumns

logger = logging.getLogger(__name__)


def run_umap_projection(
    df: pd.DataFrame,
    embedding_column: str = "SIGLIP2_EMBEDDING",
    output_lat_column: str = SchemaColumns.SIGLIP2_UMAP2D_HAVERSINE_LATITUDE,
    output_lon_column: str = SchemaColumns.SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE,
    random_state: int = 42,
    n_neighbors: int = 15,
    min_dist: float = 0.1,
) -> "umap.UMAP":
    """
    Run 2D UMAP projection on SigLIP2 embeddings with spherical embedding space.

    This function takes a DataFrame with embedding vectors and applies UMAP
    dimensionality reduction to project them into 2D coordinates on a sphere
    using Haversine distance in the output space. The resulting coordinates
    are interpreted as latitude/longitude coordinates for geographic-style analysis.

    See: https://umap-learn.readthedocs.io/en/latest/embedding_space.html#spherical-embeddings

    Missing embeddings are handled properly - rows with missing embeddings
    will have NaN values in the output coordinate columns.

    Raises:
        ValueError: If required columns are missing or if no valid embeddings found
        TypeError: If embeddings are not numpy arrays
    """
    import umap


    # Validate required columns exist
    if embedding_column not in df.columns:
        raise ValueError(f"Required column '{embedding_column}' not found in DataFrame")

    # Create result DataFrame with original data
    result_df = df

    # Initialize output columns with NaN
    result_df[output_lat_column] = np.nan
    result_df[output_lon_column] = np.nan

    # Find rows with valid (non-null) embeddings
    valid_embedding_mask = df[embedding_column].notna()

    if not valid_embedding_mask.any():
        logger.warning("No valid embeddings found, all coordinates will be NaN")
        return result_df

    # Extract valid embeddings using pandas operations
    valid_embeddings_series = df.loc[valid_embedding_mask, embedding_column]

    # Validate first embedding to check type and shape
    first_embedding = valid_embeddings_series.iloc[0]
    if not isinstance(first_embedding, np.ndarray):
        raise TypeError(f"Embeddings must be numpy arrays, got {type(first_embedding)}")
    if first_embedding.ndim != 1:
        raise ValueError(f"Embeddings must be 1D, got shape {first_embedding.shape}")

    # Stack all valid embeddings into a matrix using pandas operations
    embedding_matrix = np.stack(valid_embeddings_series.values)

    logger.info(f"Running UMAP on {len(valid_embeddings_series)} valid embeddings "
                f"with shape {embedding_matrix.shape}")

    # Configure and run UMAP
    # Ensure n_neighbors is valid (must be >= 2 and < n_samples)
    n_valid = len(valid_embeddings_series)
    effective_n_neighbors = min(n_neighbors, n_valid - 1)
    effective_n_neighbors = max(2, effective_n_neighbors)  # UMAP requires n_neighbors >= 2

    # Handle case where we have only 1 embedding (UMAP can't run)
    if n_valid == 1:
        logger.warning("Only 1 valid embedding found, assigning center coordinates")
        # Use pandas operations to assign center coordinates
        result_df.loc[valid_embedding_mask, output_lat_column] = 0.0
        result_df.loc[valid_embedding_mask, output_lon_column] = 0.0
        return result_df

    # Configure UMAP with spherical embedding space using Haversine distance
    # See: https://umap-learn.readthedocs.io/en/latest/embedding_space.html#spherical-embeddings
    sphere_mapper = umap.UMAP(
        n_components=2,
        metric="cosine",  # Input space metric (e.g., cosine for normalized embeddings)
        output_metric="haversine",  # Output space metric (haversine for spherical)
        random_state=random_state,
        n_neighbors=effective_n_neighbors,
        min_dist=min_dist,
    ).fit(embedding_matrix)

    # Go through the spherical coordinates and convert to normalized latitude and longitude.
    x = np.sin(sphere_mapper.embedding_[:, 0]) * np.cos(sphere_mapper.embedding_[:, 1])
    y = np.sin(sphere_mapper.embedding_[:, 0]) * np.sin(sphere_mapper.embedding_[:, 1])
    z = np.cos(sphere_mapper.embedding_[:, 0])

    latitude = np.degrees(-np.arccos(z)) + 90
    longitude = np.degrees(np.arctan2(x, y))

    result_df.loc[valid_embedding_mask, output_lat_column] = latitude
    result_df.loc[valid_embedding_mask, output_lon_column] = longitude

    result_df[output_lat_column] = result_df[output_lat_column].clip(lower=-90, upper=90)
    result_df[output_lon_column] = result_df[output_lon_column].clip(lower=-180, upper=180)

    logger.info(f"UMAP projection completed. Valid coordinates assigned to {n_valid} rows")

    return sphere_mapper
