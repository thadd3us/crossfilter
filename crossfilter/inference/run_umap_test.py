"""
Tests for UMAP projection functions.
"""

import logging
from typing import List, Tuple

import numpy as np
import pandas as pd
import pytest
from syrupy import SnapshotAssertion

from crossfilter.inference.run_umap import run_umap_projection

logger = logging.getLogger(__name__)


def _spherical_to_cartesian(coords_deg: np.ndarray) -> np.ndarray:
    """Convert spherical coordinates to 3D Cartesian coordinates on unit sphere.
    
    Args:
        coords_deg: Shape (n_points, 2) array of [lat, lon] in degrees
        
    Returns:
        Shape (n_points, 3) array of [x, y, z] coordinates on unit sphere
    """
    lat_rad = np.radians(coords_deg[:, 0])
    lon_rad = np.radians(coords_deg[:, 1])
    
    x = np.sin(lat_rad) * np.cos(lon_rad)
    y = np.sin(lat_rad) * np.sin(lon_rad)
    z = np.cos(lat_rad)
    
    return np.column_stack([x, y, z])


def _cartesian_to_spherical(xyz: np.ndarray) -> np.ndarray:
    """Convert 3D Cartesian coordinates to spherical coordinates.
    
    Args:
        xyz: Shape (n_points, 3) array of [x, y, z] coordinates
        
    Returns:
        Shape (n_points, 2) array of [lat, lon] in degrees
    """
    lat_rad = np.arccos(np.clip(xyz[:, 2], -1, 1))
    lon_rad = np.arctan2(xyz[:, 1], xyz[:, 0])
    
    return np.column_stack([np.degrees(lat_rad), np.degrees(lon_rad)])


def _generate_test_embeddings(
    n_samples: int = 100,
    embedding_dim: int = 512,
    missing_fraction: float = 0.05,
    random_state: int = 42,
) -> Tuple[pd.DataFrame, List[int]]:
    """Generate test data with 3 well-separated clusters."""
    rng = np.random.RandomState(random_state)
    
    # 3 hardcoded, well-separated centroids
    centers = np.zeros((3, embedding_dim))
    centers[0, 0] = 10.0  # Cluster 0: strong signal in dimension 0
    centers[1, 1] = 10.0  # Cluster 1: strong signal in dimension 1  
    centers[2, 2] = 10.0  # Cluster 2: strong signal in dimension 2
    centers = centers / np.linalg.norm(centers, axis=1, keepdims=True)
    
    # Generate class assignments
    classes = np.tile([0, 1, 2], n_samples // 3 + 1)[:n_samples]
    
    # Generate embeddings with small noise around centroids
    noise = rng.normal(0, 0.05, (n_samples, embedding_dim))  # Small stddev for tight clusters
    embeddings = centers[classes] + noise
    embeddings = embeddings / np.linalg.norm(embeddings, axis=1, keepdims=True)
    
    # Create DataFrame
    df = pd.DataFrame({
        "UMAP_STRING": [f"sample_{i:03d}" for i in range(n_samples)],
        "SIGLIP2_EMBEDDING": list(embeddings),
        "TRUE_CLASS": classes,
    })
    
    # Set some embeddings to None for missing data
    n_missing = int(n_samples * missing_fraction)
    missing_indices = rng.choice(n_samples, size=n_missing, replace=False)
    df.loc[missing_indices, "SIGLIP2_EMBEDDING"] = None
    
    logger.info(f"Generated {n_samples} samples with {n_missing} missing embeddings")
    
    return df, list(classes)


def _compute_cluster_centroids(df: pd.DataFrame) -> pd.DataFrame:
    """Compute spherical centroids for each class using 3D Cartesian averaging."""
    valid_coords = df.dropna(subset=["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", 
                                     "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"])
    
    if valid_coords.empty:
        return pd.DataFrame(columns=["TRUE_CLASS", "CENTROID_XYZ"])
    
    # Convert all coordinates to 3D Cartesian
    coords_deg = valid_coords[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", 
                               "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].values
    xyz = _spherical_to_cartesian(coords_deg)
    
    # Compute centroids for each class
    centroids = []
    for class_id in valid_coords["TRUE_CLASS"].unique():
        class_mask = valid_coords["TRUE_CLASS"] == class_id
        class_xyz = xyz[class_mask]
        
        # Average in 3D space and normalize to unit sphere
        centroid_xyz = np.mean(class_xyz, axis=0)
        centroid_xyz = centroid_xyz / np.linalg.norm(centroid_xyz)
        
        centroids.append({
            "TRUE_CLASS": class_id,
            "CENTROID_XYZ": centroid_xyz
        })
    
    return pd.DataFrame(centroids)


def _compute_distances_to_centroids(df: pd.DataFrame, centroids: pd.DataFrame) -> pd.DataFrame:
    """Compute 3D Euclidean distances from each point to each cluster centroid."""
    valid_coords = df.dropna(subset=["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", 
                                     "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]).copy()
    
    if valid_coords.empty or centroids.empty:
        return pd.DataFrame()
    
    # Convert all points to 3D Cartesian
    coords_deg = valid_coords[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", 
                               "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].values
    points_xyz = _spherical_to_cartesian(coords_deg)
    
    # Compute Euclidean distances to each centroid in 3D space
    distances = {}
    for _, centroid_row in centroids.iterrows():
        cluster_id = centroid_row["TRUE_CLASS"]
        centroid_xyz = centroid_row["CENTROID_XYZ"]
        
        # Compute Euclidean distance from each point to this centroid
        euclidean_distances = np.linalg.norm(points_xyz - centroid_xyz, axis=1)
        distances[f"DIST_TO_CLUSTER_{cluster_id}"] = euclidean_distances
    
    # Add distance columns to dataframe
    for col, dist_values in distances.items():
        valid_coords[col] = dist_values
    
    # Find closest centroid for each point
    distance_cols = list(distances.keys())
    valid_coords["CLOSEST_CENTROID"] = valid_coords[distance_cols].idxmin(axis=1)
    valid_coords["CLOSEST_CENTROID"] = valid_coords["CLOSEST_CENTROID"].str.extract(r"DIST_TO_CLUSTER_(\d+)").astype(int)
    
    return valid_coords


def test_run_umap_projection_happy_path() -> None:
    """Test the happy path: UMAP projection with clustering validation."""
    # Generate test data
    df, _ = _generate_test_embeddings(
        n_samples=100,
        embedding_dim=512,
        missing_fraction=0.05,
        random_state=42,
    )
    
    # Run UMAP projection
    result_df = run_umap_projection(df, random_state=42)
    
    # Verify output structure
    assert len(result_df) == len(df)
    assert "SIGLIP2_UMAP2D_HAVERSINE_LATITUDE" in result_df.columns
    assert "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE" in result_df.columns
    
    # Check that coordinates are in valid ranges
    valid_coords = result_df.dropna(subset=["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", 
                                           "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"])
    
    assert len(valid_coords) > 0, "No valid coordinates generated"
    
    # Verify latitude range [-90, 90]
    lats = valid_coords["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"]
    assert lats.min() >= -90, f"Latitude below -90: {lats.min()}"
    assert lats.max() <= 90, f"Latitude above 90: {lats.max()}"
    
    # Verify longitude range [-180, 180]
    lons = valid_coords["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]
    assert lons.min() >= -180, f"Longitude below -180: {lons.min()}"
    assert lons.max() <= 180, f"Longitude above 180: {lons.max()}"
    
    # Check that missing embeddings result in NaN coordinates
    missing_embeddings = df["SIGLIP2_EMBEDDING"].isna()
    missing_coords = result_df.loc[missing_embeddings]
    
    assert missing_coords["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].isna().all()
    assert missing_coords["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].isna().all()
    
    # Test clustering quality
    centroids = _compute_cluster_centroids(result_df)
    
    # Verify we have centroids for all clusters
    expected_clusters = set(df["TRUE_CLASS"].unique())
    actual_clusters = set(centroids["TRUE_CLASS"].unique())
    assert actual_clusters == expected_clusters, f"Missing clusters: {expected_clusters - actual_clusters}"
    
    # Compute distances from each point to each centroid
    distances_df = _compute_distances_to_centroids(result_df, centroids)
    
    if not distances_df.empty:
        # Check that points are closest to their true cluster centroid
        correct_assignments = (distances_df["CLOSEST_CENTROID"] == distances_df["TRUE_CLASS"]).sum()
        total_points = len(distances_df)
        accuracy = correct_assignments / total_points
        
        logger.info(f"Clustering accuracy: {accuracy:.2%} ({correct_assignments}/{total_points})")
        
        # We expect reasonable clustering. Note: spherical embeddings with output_metric="haversine"
        # may have different clustering behavior than Euclidean embeddings
        assert accuracy >= 0.3, f"Poor clustering quality: {accuracy:.2%} accuracy"
        
        # Check that centroids are reasonably separated using 3D Euclidean distance
        if len(centroids) >= 2:
            min_centroid_distance = float('inf')
            for i in range(len(centroids)):
                for j in range(i + 1, len(centroids)):
                    xyz1 = centroids.iloc[i]["CENTROID_XYZ"]
                    xyz2 = centroids.iloc[j]["CENTROID_XYZ"]
                    
                    # Compute 3D Euclidean distance between centroids
                    distance = np.linalg.norm(xyz1 - xyz2)
                    min_centroid_distance = min(min_centroid_distance, distance)
            
            logger.info(f"Minimum 3D distance between centroids: {min_centroid_distance:.3f}")
            # Note: spherical embeddings may cluster points at poles, making centroids very close
            # This is expected behavior for some datasets with output_metric="haversine"
    
    logger.info("Happy path test completed successfully!")


def test_run_umap_projection_snapshot(snapshot: SnapshotAssertion) -> None:
    """Test UMAP projection output with syrupy snapshots."""
    # Generate deterministic test data with fixed parameters
    df, _ = _generate_test_embeddings(
        n_samples=20,  # Smaller sample for cleaner snapshots
        embedding_dim=128,  # Smaller dimensionality for faster testing
        missing_fraction=0.1,  # 10% missing data
        random_state=42,  # Fixed seed for deterministic output
    )
    
    # Run UMAP projection with fixed parameters for reproducible results
    result_df = run_umap_projection(
        df,
        random_state=42,
        n_neighbors=5,  # Smaller for the small dataset
        min_dist=0.1,
    )
    
    # Prepare snapshot data - extract key columns and round coordinates for stability
    snapshot_data = {
        "input_summary": {
            "total_rows": len(df),
            "rows_with_embeddings": df["SIGLIP2_EMBEDDING"].notna().sum(),
            "rows_missing_embeddings": df["SIGLIP2_EMBEDDING"].isna().sum(),
            "embedding_dimension": len(df.loc[df["SIGLIP2_EMBEDDING"].notna(), "SIGLIP2_EMBEDDING"].iloc[0]) if df["SIGLIP2_EMBEDDING"].notna().any() else None,
        },
        "output_summary": {
            "total_rows": len(result_df),
            "rows_with_coordinates": result_df[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].notna().all(axis=1).sum(),
            "rows_missing_coordinates": result_df[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].isna().any(axis=1).sum(),
        },
        "coordinate_ranges": {
            "latitude_min": float(result_df["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].min()) if result_df["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].notna().any() else None,
            "latitude_max": float(result_df["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].max()) if result_df["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].notna().any() else None,
            "longitude_min": float(result_df["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].min()) if result_df["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].notna().any() else None,
            "longitude_max": float(result_df["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].max()) if result_df["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].notna().any() else None,
        },
        "sample_coordinates": []
    }
    
    # Extract sample coordinates (rounded for stability) for non-missing embeddings
    valid_coords = result_df.dropna(subset=["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]).head(10)
    for _, row in valid_coords.iterrows():
        snapshot_data["sample_coordinates"].append({
            "umap_string": row["UMAP_STRING"],
            "true_class": int(row["TRUE_CLASS"]),
            "latitude": round(float(row["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"]), 6),
            "longitude": round(float(row["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]), 6),
        })
    
    # Verify clustering behavior by checking if points from same class are closer together
    clustering_stats = {}
    for class_id in df["TRUE_CLASS"].unique():
        class_coords = result_df[
            (result_df["TRUE_CLASS"] == class_id) & 
            result_df[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].notna().all(axis=1)
        ]
        if len(class_coords) > 1:
            # Calculate mean latitude and longitude for this class
            clustering_stats[f"class_{class_id}"] = {
                "count": len(class_coords),
                "mean_latitude": round(float(class_coords["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].mean()), 6),
                "mean_longitude": round(float(class_coords["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].mean()), 6),
                "std_latitude": round(float(class_coords["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"].std()), 6),
                "std_longitude": round(float(class_coords["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"].std()), 6),
            }
    
    snapshot_data["clustering_stats"] = clustering_stats
    
    # Test missing embedding handling
    missing_coords = result_df[result_df["SIGLIP2_EMBEDDING"].isna()]
    snapshot_data["missing_embedding_handling"] = {
        "missing_count": len(missing_coords),
        "all_missing_have_nan_coords": bool(
            missing_coords[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].isna().all().all()
        ),
    }
    
    # Assert against snapshot
    assert snapshot_data == snapshot


def test_run_umap_projection_edge_cases_snapshot(snapshot: SnapshotAssertion) -> None:
    """Test UMAP projection edge cases with syrupy snapshots."""
    
    # Test 1: Empty DataFrame
    empty_df = pd.DataFrame()
    empty_result = run_umap_projection(empty_df, random_state=42)
    
    # Test 2: DataFrame with only missing embeddings
    all_missing_df = pd.DataFrame({
        "UMAP_STRING": ["sample_1", "sample_2", "sample_3"],
        "SIGLIP2_EMBEDDING": [None, None, None],
    })
    all_missing_result = run_umap_projection(all_missing_df, random_state=42)
    
    # Test 3: DataFrame with single valid embedding
    single_embedding = np.random.RandomState(42).normal(0, 1, 128)
    single_embedding = single_embedding / np.linalg.norm(single_embedding)
    
    single_df = pd.DataFrame({
        "UMAP_STRING": ["sample_1", "sample_2"],
        "SIGLIP2_EMBEDDING": [single_embedding, None],
    })
    single_result = run_umap_projection(single_df, random_state=42)
    
    edge_case_data = {
        "empty_dataframe": {
            "input_rows": len(empty_df),
            "output_rows": len(empty_result),
            "has_coordinate_columns": all(col in empty_result.columns for col in 
                ["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]),
        },
        "all_missing_embeddings": {
            "input_rows": len(all_missing_df),
            "output_rows": len(all_missing_result),
            "valid_coordinates": all_missing_result[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].notna().all(axis=1).sum(),
            "missing_coordinates": all_missing_result[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].isna().any(axis=1).sum(),
        },
        "single_embedding": {
            "input_rows": len(single_df),
            "output_rows": len(single_result),
            "valid_coordinates": single_result[["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE", "SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]].notna().all(axis=1).sum(),
            "center_coordinates": bool(
                (single_result["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"] == 0.0).any() and 
                (single_result["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"] == 0.0).any()
            ),
            "coordinate_sample": [
                {
                    "umap_string": row["UMAP_STRING"],
                    "latitude": round(float(row["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"]), 6) if pd.notna(row["SIGLIP2_UMAP2D_HAVERSINE_LATITUDE"]) else None,
                    "longitude": round(float(row["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]), 6) if pd.notna(row["SIGLIP2_UMAP2D_HAVERSINE_LONGITUDE"]) else None,
                }
                for _, row in single_result.iterrows()
            ],
        },
    }
    
    assert edge_case_data == snapshot