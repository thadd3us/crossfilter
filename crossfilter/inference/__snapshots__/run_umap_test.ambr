# serializer version: 1
# name: test_run_umap_projection_edge_cases_snapshot
  dict({
    'all_missing_embeddings': dict({
      'input_rows': 3,
      'missing_coordinates': 3,
      'output_rows': 3,
      'valid_coordinates': 0,
    }),
    'empty_dataframe': dict({
      'has_coordinate_columns': True,
      'input_rows': 0,
      'output_rows': 0,
    }),
    'single_embedding': dict({
      'center_coordinates': True,
      'coordinate_sample': list([
        dict({
          'latitude': 0.0,
          'longitude': 0.0,
          'umap_string': 'sample_1',
        }),
        dict({
          'latitude': None,
          'longitude': None,
          'umap_string': 'sample_2',
        }),
      ]),
      'input_rows': 2,
      'output_rows': 2,
      'valid_coordinates': 1,
    }),
  })
# ---
# name: test_run_umap_projection_snapshot
  dict({
    'clustering_stats': dict({
      'class_0': dict({
        'count': 6,
        'mean_latitude': -8.330429,
        'mean_longitude': 43.739677,
        'std_latitude': 48.733766,
        'std_longitude': 93.677404,
      }),
      'class_1': dict({
        'count': 7,
        'mean_latitude': -5.769924,
        'mean_longitude': 10.55033,
        'std_latitude': 42.93811,
        'std_longitude': 96.201742,
      }),
      'class_2': dict({
        'count': 5,
        'mean_latitude': 3.175984,
        'mean_longitude': 45.943509,
        'std_latitude': 42.818948,
        'std_longitude': 71.123782,
      }),
    }),
    'coordinate_ranges': dict({
      'latitude_max': 90.0,
      'latitude_min': -90.0,
      'longitude_max': 180.0,
      'longitude_min': -180.0,
    }),
    'input_summary': dict({
      'embedding_dimension': 128,
      'rows_missing_embeddings': 2,
      'rows_with_embeddings': 18,
      'total_rows': 20,
    }),
    'missing_embedding_handling': dict({
      'all_missing_have_nan_coords': True,
      'missing_count': 2,
    }),
    'output_summary': dict({
      'rows_missing_coordinates': 2,
      'rows_with_coordinates': 18,
      'total_rows': 20,
    }),
    'sample_coordinates': list([
      dict({
        'latitude': 90.0,
        'longitude': -31.052109,
        'true_class': 0,
        'umap_string': 'sample_000',
      }),
      dict({
        'latitude': 54.306122,
        'longitude': 105.612396,
        'true_class': 1,
        'umap_string': 'sample_001',
      }),
      dict({
        'latitude': 25.496391,
        'longitude': 135.964813,
        'true_class': 2,
        'umap_string': 'sample_002',
      }),
      dict({
        'latitude': 6.391304,
        'longitude': 43.591858,
        'true_class': 1,
        'umap_string': 'sample_004',
      }),
      dict({
        'latitude': -36.841385,
        'longitude': 138.447052,
        'true_class': 0,
        'umap_string': 'sample_006',
      }),
      dict({
        'latitude': -5.512314,
        'longitude': 3.024902,
        'true_class': 1,
        'umap_string': 'sample_007',
      }),
      dict({
        'latitude': 55.602005,
        'longitude': -50.934052,
        'true_class': 2,
        'umap_string': 'sample_008',
      }),
      dict({
        'latitude': -36.259979,
        'longitude': -28.603043,
        'true_class': 0,
        'umap_string': 'sample_009',
      }),
      dict({
        'latitude': 3.0569,
        'longitude': -180.0,
        'true_class': 1,
        'umap_string': 'sample_010',
      }),
      dict({
        'latitude': 14.296928,
        'longitude': 90.340149,
        'true_class': 2,
        'umap_string': 'sample_011',
      }),
    ]),
  })
# ---
